merkle_tree/src/internal.rs

```diff
@@ -131,52 +128,76 @@ impl<T: NodeValue> MerkleCommitment<T> for MerkleTreeCommitment<T> {
+/// Merkle path.
+#[derive(
+    Clone, Debug, Hash, PartialEq, Eq, PartialOrd, Ord, CanonicalSerialize, CanonicalDeserialize,
+)]
+#[tagged("MERKLE_PROOF")]
+pub struct MerkleTreeProof<T: NodeValue>(pub Vec<Vec<T>>);
+
+impl<T: NodeValue> super::MerkleProof<T> for MerkleTreeProof<T> {
+    /// Expected height of the Merkle tree.
+    fn height(&self) -> usize {
+        self.0.len()
+    }
+
+    /// Return all values of siblings of this Merkle path
+    fn path_values(&self) -> &[Vec<T>] {
+        &self.0
+    }
 }
 
-impl<E, I, T, const ARITY: usize> MerkleProof<E, I, T, ARITY>
+/// Verify a merkle proof
+/// * `commitment` - a merkle tree commitment
+/// * `pos` - zero-based index of the leaf in the tree
+/// * `element` - the leaf value, None if verifying a non-membership proof
+/// * `proof` - a membership proof for `element` at given `pos`
+/// * `returns` - Ok(true) if the proof is accepted, Ok(false) if not. Err() if
+///   the proof is not well structured, E.g. not for this merkle tree.
+pub(crate) fn verify_merkle_proof<E, H, I, const ARITY: usize, T>(
+    commitment: &MerkleTreeCommitment<T>,
+    pos: &I,
+    element: Option<&E>,
+    proof: &[Vec<T>],
+) -> Result<VerificationResult, MerkleTreeError>
 where
     E: Element,
-    I: Index,
+    I: Index + ToTraversalPath<ARITY>,
     T: NodeValue,
+    H: DigestAlgorithm<E, I, T>,
 {
-    /// Return the height of this proof.
-    pub fn tree_height(&self) -> usize {
-        self.proof.len()
-    }
-
-    /// Form a `MerkleProof` from a given index and Merkle path.
-    pub fn new(pos: I, proof: MerklePath<E, I, T>) -> Self {
-        MerkleProof { pos, proof }
-    }
-
-    /// Return the index of this `MerkleProof`.
-    pub fn index(&self) -> &I {
-        &self.pos
-    }
-
-    /// Return the element associated with this `MerkleProof`. None if it's a
-    /// non-membership proof.
-    pub fn elem(&self) -> Option<&E> {
-        match self.proof.first() {
-            Some(MerkleNode::Leaf { elem, .. }) => Some(elem),
-            _ => None,
+    if proof.len() == commitment.height() {
+        let init = if let Some(elem) = element {
+            H::digest_leaf(pos, elem)?
+        } else {
+            T::default()
+        };
+        let computed_root = pos
+            .to_traversal_path(commitment.height())
+            .iter()
+            .zip(proof.iter())
+            .try_fold(
+                init,
+                |val, (branch, values)| -> Result<T, MerkleTreeError> {
+                    if values.len() == 0 {
+                        Ok(T::default())
+                    } else {
+                        let mut data = values.clone();
+                        data.insert(*branch, val);
+                        H::digest(&data)
+                    }
+                },
+            )?;
+        if computed_root == commitment.digest() {
+            Ok(Ok(()))
+        } else {
+            Ok(Err(()))
         }
+    } else {
```

Comment:
nit: in order to avoid excessive indentation consider coding like this:

```rust
if (error condition) {
  return error;
}

// remaining code
```

Reply:
Chaged in [11e0f4d](https://github.com/EspressoSystems/jellyfish/pull/692/commits/11e0f4debbc8142925649d187f24fd0703c95934)